File: modules/foreground-service/src/ForegroundServiceModule.ts
import { NativeModule, requireNativeModule } from 'expo';

import type { ForegroundServiceModuleEvents } from './ForegroundService.types';

declare class ForegroundServiceModule extends NativeModule<ForegroundServiceModuleEvents> {
  start(title: string, text: string): void;
  stop(): void;
}

export default requireNativeModule<ForegroundServiceModule>('ForegroundService');

File: modules/foreground-service/android/src/main/java/expo/modules/foregroundservice/ForegroundServiceModule.kt
package expo.modules.foregroundservice

import android.content.Intent
import androidx.core.content.ContextCompat
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ForegroundServiceModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ForegroundService")

    Function("start") { title: String, text: String ->
      appContext.reactContext?.let { context ->
        val intent = Intent(context, SyncForegroundService::class.java).apply {
          putExtra(SyncForegroundService.EXTRA_TITLE, title)
          putExtra(SyncForegroundService.EXTRA_TEXT, text)
        }
        ContextCompat.startForegroundService(context, intent)
      }
    }

    Function("stop") {
      appContext.reactContext?.let { context ->
        context.stopService(Intent(context, SyncForegroundService::class.java))
      }
    }
  }
}
